usb/ehci/ehci.c

```diff
@@ -0,0 +1,753 @@
+			;
+	}
+}
+
+
+static void ehci_qtdsFree(struct qtd_node **head)
+{
+	struct qtd_node *q;
+
+	while ((q = *head) != NULL) {
+		LIST_REMOVE(head, q);
+		usb_free(q->qtd, sizeof(struct qtd));
+		free(q);
+	}
+}
+
+static void ehci_qtdsDeactivate(struct qtd_node *qtds)
+{
+	struct qtd_node *e = qtds;
+
+	if (e != NULL) {
+		do {
+			e->qtd->active = 0;
+		} while ((e = e->next) != qtds);
+	}
+}
+
+
+void ehci_unlinkQhAsync(hcd_t *hcd, struct qh_node *qh)
+{
+	ehci_t *ehci = (ehci_t *)hcd->priv;
+
+	if (qh->qh->headOfReclamation)
+		qh->next->qh->headOfReclamation = 1;
+
+	if (qh->qh == qh->next->qh) {
+		ehci->asyncList = NULL;
+		*(hcd->base + usbcmd) &= ~USBCMD_ASE;
+		while (*(hcd->base + usbsts) & USBSTS_AS)
+			;
+		*(hcd->base + asynclistaddr) = 1;
+	}
+	else if (qh == ehci->asyncList) {
+		ehci->asyncList = qh->next;
+		*(hcd->base + asynclistaddr) = va2pa(qh->next->qh);
+	}
+
+	qh->prev->qh->horizontal = qh->qh->horizontal;
+	LIST_REMOVE(&ehci->asyncList, qh);
+}
+
+void ehci_unlinkQhPeriodic(hcd_t *hcd, struct qh_node *qh)
+{
+	ehci_t *ehci = (ehci_t *)hcd->priv;
+	struct qh_node *tmp;
+	int i;
+
+	/* TODO: do we have to stop the periodic queue? */
+	for (i = 0; i < EHCI_PERIODIC_SIZE; i++) {
+		/* Count Qhs linked to this periodic index */
+		tmp = ehci->periodicNodes[i];
+
+		if (tmp == qh) {
+			if (qh->next != NULL) {
+				ehci->periodicList[i].pointer = va2pa(qh->next->qh) >> 5;
+			}
+			else {
+				ehci->periodicList[i].pointer = 0;
+				ehci->periodicList[i].terminate = 1;
+			}
+			ehci->periodicNodes[i] = qh->next;
+		}
+		else {
+			while (tmp != NULL && tmp->next != qh)
+				tmp = tmp->next;
+
+			if (tmp != NULL && tmp->next == qh) {
+				tmp->next = qh->next;
+				if (tmp->next != NULL) {
+					tmp->qh->horizontal.pointer = va2pa(tmp->next->qh) >> 5;
+				}
+				else {
+					tmp->qh->horizontal.pointer = 0;
+					tmp->qh->horizontal.terminate = 1;
+				}
+			}
+		}
+	}
+}
+
+
+static int ehci_irqHandler(unsigned int n, void *data)
+{
+	hcd_t *hcd = (hcd_t *)data;
+	ehci_t *ehci = (ehci_t *)hcd->priv;
+
+	ehci->status = *(hcd->base + usbsts);
+	*(hcd->base + usbsts) = ehci->status & 0x1f;
+
+	return -!(ehci->status & 0x1f);
+}
+
+
+static void ehci_qtdsCheck(hcd_t *hcd, usb_transfer_t *t)
+{
+	struct qtd_node *qtds = (struct qtd_node *)t->hcdpriv;
+	int error = 0;
+
+	do {
+		if (qtds->qtd->transactionError || qtds->qtd->babble)
+			error++;
+
+		qtds = qtds->next;
+	} while (qtds != t->hcdpriv);
+
+	if (!qtds->prev->qtd->active || qtds->prev->qtd->halted || error > 0)
+		t->finished = 1;
+
+	t->error = error;
+	t->transferred = t->size - qtds->prev->qtd->bytesToTransfer;
+}
+
+
+static void ehci_transUpdate(hcd_t *hcd)
+{
+	struct qh_node *qh;
+	struct qtd_node *qtd;
+	usb_transfer_t *t, *n;
+	int cont;
+
+	if ((t = hcd->transfers) == NULL)
+		return;
+
+	do {
+		qh = (struct qh_node *)t->pipe->hcdpriv;
+		qtd = (struct qtd_node *)t->hcdpriv;
+		cont = 0;
+		n = t->next;
+
+		ehci_qtdsCheck(hcd, t);
+
+		if (t->finished) {
```

Comment:
This variable should be always accessed with the same mutex (usbdev_common.lock in phoenix-rtos-usb)